File: 10-Recursion-and-Backtracking/python/01_permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        result = []

        def backtrack(index):
            if index == len(nums):
                result.append(list(nums))
                return

            for i in range(index, len(nums)):
                nums[index], nums[i] = nums[i], nums[index]
                backtrack(index + 1)
                nums[index], nums[i] = nums[i], nums[index]

        backtrack(0)
        return result

File: 10-Recursion-and-Backtracking/cpp/01-permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> result;
        backtrack(nums, 0, result);
        return result;
    }

private:
    void backtrack(vector<int>& nums, int index, vector<vector<int>>& result) {
        if (index == (int)nums.size()) {
            result.push_back(nums);
            return;
        }

        for (int i = index; i < (int)nums.size(); i++) {
            swap(nums[index], nums[i]);
            backtrack(nums, index + 1, result);
            swap(nums[index], nums[i]);
        }
    }
};
